pyro/contrib/epidemiology/sir.py

```diff
@@ -212,3 +210,111 @@ def transition_bwd(self, params, prev, curr, t):
+    intervals. To preserve Markov structure (and hence tractible inference)
+    this model adds an auxiliary compartment ``O`` denoting the fully-observed
+    cumulative number of observations at each time point. At observed times
+    (when ``mask[t] == True``) ``O`` must exactly match the provided data;
+    between observed times ``O`` stochastically imputes the provided data.
+
+    :param int population: Total ``population = S + I + R``.
+    :param float recovery_time: Mean recovery time (duration in state
+        ``I``). Must be greater than 1.
+    :param iterable data: Time series of **cumulative** observed infections.
+        Whenever ``mask[t] == True``, ``data[t]`` corresponds to an
+        observation; otherwise ``data[t]`` can be arbitrary, e.g. NAN.
+    :param iterable mask: Boolean time series denoting whether an observation
+        is made at each time step. Should satisfy ``len(mask) == len(data)``.
+    """
+
+    def __init__(self, population, recovery_time, data, mask):
+        assert len(data) == len(mask)
+        duration = len(data)
+        compartments = ("S", "I", "O")  # O is auxiliary, R is implicit.
+        super().__init__(compartments, duration, population)
+
+        assert isinstance(recovery_time, float)
+        assert recovery_time > 1
+        self.recovery_time = recovery_time
+
+        self.data = data
+        self.mask = mask
+
+    series = ("S2I", "I2R", "S2O", "obs")
+    full_mass = [("R0", "rho")]
+
+    def global_model(self):
+        tau = self.recovery_time
+        R0 = pyro.sample("R0", dist.LogNormal(0., 1.))
+        rho = pyro.sample("rho", dist.Uniform(0, 1))
+        return R0, tau, rho
+
+    def initialize(self, params):
+        # Start with a single infection.
+        return {"S": self.population - 1, "I": 1, "O": 0}
+
+    def transition_fwd(self, params, state, t):
+        R0, tau, rho = params
+
+        # Sample flows between compartments.
+        S2I = pyro.sample("S2I_{}".format(t),
+                          infection_dist(individual_rate=R0 / tau,
+                                         num_susceptible=state["S"],
+                                         num_infectious=state["I"],
+                                         population=self.population))
+        I2R = pyro.sample("I2R_{}".format(t),
+                          dist.Binomial(state["I"], 1 / tau))
+        S2O = pyro.sample("S2O_{}".format(t),
+                          dist.ExtendedBinomial(S2I, rho))
+
+        # Update compartments with flows.
+        state["S"] = state["S"] - S2I
+        state["I"] = state["I"] + S2I - I2R
+        state["O"] = state["O"] + S2O
+
+        # Condition on cumulative observations.
+        mask_t = self.mask[t] if t < self.duration else False
+        data_t = self.data[t] if t < self.duration else None
+        pyro.sample("obs_{}".format(t),
+                    dist.Delta(state["O"]).mask(mask_t),
```

Comment:
why is the auxiliary necessary? or does this just play better with the structure of `CompartmentalModel`?

Reply:
tl;dr The auxiliary variable is needed to preserve Markov structure.

The observations in this model are aggregated over intervals: `obs=S2I[t_prev+1:t_curr+1].sum()` where `t_prev` is the time of the last observation and `t_curr` is the time of the current observation. In our enumeration strategy, this would couple all `t_curr-t_prev`-many enumeration variables, growing exponentially in the number of variables. While the non-parallel-scan enumeration strategy could handle this, it would be prohibitively expensive, and would not allow e.g. large gaps in sensor data (as e.g. when a government shuts down or runs out of tests). The trick we're using is to add an auxiliary variable for the entire cumulative observation trajectory (with the same likelihood as in the usual SIR models), and then `Delta`-clamp that auxiliary to the true observations at a few sparse time steps. This makes more work for HMC adds one enumeration variable per time step and increases the complexity of variable elimination by a constant factor of `Q**2`, but crucially this factor is independent of gap size.

I had been struggling with this issue for a while since Lucy's model simulates 4 times per day but is observed only once. The only alternative I could see was to do parallel-scan variable elimination where each `DiscreteHMM` state covered the joint distribution over an entire day (four time steps), resulting in complexity `Q**(2 * 4 * 2)` for an SIR model or `Q**(3 * 4 * 2)` for an SEIR model.

Reply:
thanks for the explanation. to clarify though: if all you had was occasional missing data you wouldn't need this construction. this is really for the cumulative case

Reply:
Correct. It appears the cumulative case is more common in epidemiology.